cap-ci refactoring, modularization, and fixing job dependencies

HartS opened the pull request:
Cap-ci templating and refactoring:
    - Rename 'clean' to clearer 'destroy-kubecf'
    - Move embedded scripts into modular templates
    - Add ordered list of jobs to config
    - Use config via --template instead of --datasource
    - Rename 'options' to 'availabilities'

    Improve job dependencies with modularization
    - Move jobs definitions into module templates
    - Generate jobs from ordered list. This makes it possible to name the previous
      job dependencies correctly with the 'passed:' get step key

Commit:
cap-ci templating and refactoring:
- Rename 'clean' to clearer 'destroy-kubecf'
- Move embedded scripts into modular templates
- Add ordered list of jobs to config
- Use config via --template instead of --datasource
- Rename 'options' to 'availabilities'

Commit:
Improve job dependencies with modularization
- Move jobs definitions into module templates
- Generate jobs from ordered list. This makes it possible to name the previous
  job dependencies correctly with the 'passed:' get step key

Commit:
Remove trailing whitespace

bikramnehra commented on cap-ci/config.yaml.tmpl:
```diff
@@ -6,13 +7,22 @@ jobs:
   gke: true
   eks: false
-options:
+availabilities:
```
Did we ever figure out if we need `all` availability?

HartS replied:
I still don't think so, but I left it the way it was

bikramnehra commented on cap-ci/jobs/cf-acceptance-tests-brain.tmpl:
```diff
@@ -0,0 +1,51 @@
+{{ define "jobs_cf-acceptance-tests-brain" }}
```
Is there a reason why we can't simply use `cf-acceptance-tests-brain`? Its pretty clear since its present in jobs directory. Same for other job names.

HartS replied:
Yes, because when you template a resource, you want to prefix it with the type of resource. In this PR for example, you'd have a collision with `scripts_stratos` and `jobs_stratos` if you took away the prefix

bikramnehra commented on cap-ci/pipeline.yaml.tmpl:
```diff
@@ -0,0 +1,114 @@
+- {{ $job }}
+{{ end }}{{ end }}{{ end }}
+{{ $jobs_enabled := (tmpl.Exec "jobs_ordered" $config |  data.YAMLArray ) }}
+groups:
```
Let's add a newline before `groups`.

HartS replied:
There should be a couple in the output already

bikramnehra commented on cap-ci/scripts/deploy.tmpl:
```diff
@@ -0,0 +1,11 @@
+{{ define "scripts_deploy" }}
+# Attention: PWD = ___/catapult here
```
I remember you mentioning about `gomplates'` comments, is that something which we can incorporate now?

HartS replied:
Yes, I didn't do that, but I can

HartS replied:
✅

bikramnehra commented on cap-ci/scripts/deploy.tmpl:
```diff
@@ -0,0 +1,11 @@
+{{ define "scripts_deploy" }}
```
same comment as for jobs. I am wondering if its needed by `gomplate`?

HartS replied:
Not by gomplate, but it's good practice to prefix with the type of resource; see above

bikramnehra commented on cap-ci/pipeline.yaml.tmpl:
```diff
@@ -0,0 +1,114 @@
+{{- range $backend, $enabled :=  $config.backends }}{{ if $enabled }}
+{{- range $avail, $enabled := $config.availabilities }}{{ if $enabled }}
+{{- range $scheduler, $enabled := $config.schedulers }}{{ if $enabled }}
+{{- $context := $context | merge (dict "position" $job_index "backend" $backend "avail" $avail "scheduler" $scheduler) -}}
```
:+1:

bikramnehra commented on cap-ci/config.yaml.tmpl:
```diff
@@ -6,13 +7,22 @@ jobs:
-  clean: true
+  destroy-kubecf: true
+jobs_ordered:
+- deploy
```
In light of `destroy-kubecf`, can we also rename this to `deploy-kubecf`?

HartS replied:
✅

bikramnehra commented on cap-ci/config.yaml.tmpl:
```diff
@@ -6,13 +7,22 @@ jobs:
+- cf-acceptance-tests-brain
+- cf-acceptance-tests
+- sync-integration-tests
+- upgrade
```
Also, this can be `upgrade-kubecf`.

HartS replied:
✅

bikramnehra commented on cap-ci/create_cap-release_pipeline.sh:
```diff
@@ -38,4 +38,7 @@ fly_args=(
 
-fly "${fly_args[@]}" --config <(gomplate --verbose --datasource config.yaml --file pipeline.template)
+# space-separated paths to template files and directories which contain template files
+TEMPLATE_PATHS="config.yaml.tmpl scripts"
```
don't we need `jobs` directory too here?

HartS replied:
Yes, thanks for pointing that out; I removed it when I made the first commit, forgot to add it back

HartS replied:
✅

bikramnehra reviewed (changes requested):
In addition to comments, have we deployed an instance of pipeline with these changes?

Commit:
Address PR feedback
- Add `jobs/` path back into TEMPLATE_PATHS in pipeline deploy script
- rename `deploy` to `deploy-kubecf`, `upgrade` to `upgrade-kubecf`
- change scripts comment style from bash to gomplate

bikramnehra reviewed (approved):
LGTM.
